SpinCore_pp/ppg/generic.py

```diff
@@ -102,12 +103,20 @@ def generic(
     """
     tx_phases = r_[0.0, 90.0, 180.0, 270.0]
     # {{{ pull info about phase cycling and echos from the ppg_list
-    # {{{ tuples with 4 elements are pulses, where the 4th element is the phase cycle
-    all_ppg_arrays = [j[3] for j in ppg_list if len(j)>3]
-    nPhaseSteps = prod([len(j) for j in all_ppg_arrays])
-    # }}}
-    # {{{ for this to work, the loop label for echoes must be called "echo_label"
-    nEchoes = [j[2]+1 for j in ppg_list if len(j)>2 and j[0] == 'marker' and j[1] == 'echo_label']
+    # {{{ tuples with 4 elements are pulses, where the 4th element is the phase cycle and tuples containing marker and echo label are cycled by nEchoes
+    phcyc_lens = {}
+    nEchoes = 1 #assume nEchoes is 1 before going through the ppg list
+    for thiselem in ppg_list:
+        if len(thiselem) > 3: #tuples with more than 3 are phase cycling commands
+            phcyc_lens[thiselem[2]] = len(thiselem[3]) #set the name of the ph and length
+        if len(thiselem)>2 and thiselem[0] == 'marker' and thiselem[1] == 'echo_label':  
+            # if there is a tuple that calls marker AND echo label we assume this is a cpmg or
+            # ppg with multiple echoes so we reset the nEchoes to the value indicated in the 
+            # ppg list
+            nEchoes = thiselem[2]+1
+    ph_lens = list(phcyc_lens.values()) #list of phase cycling lengths
+    nPhaseSteps = prod(ph_lens) #total number of phase steps in ppg
+    print(nEchoes)
```

Comment:
Just roll back to the earlier code.  If there is a problem with the earlier code, we can discuss.

Reply:
This will not work for a generic hahn echo since there is no call to marker for a regular hahn echo - probably won't work for our split IR either